server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/factory/DeployedCacheStoreFactory.java

```diff
@@ -18,49 +21,68 @@
  */
 public class DeployedCacheStoreFactory implements CacheStoreFactory {
 
-   private Map<String, DeployedCacheStoreMetadata> deployedCacheStores = Collections.synchronizedMap(new HashMap<String, DeployedCacheStoreMetadata>());
+   private static final int TIMEOUT_SECONDS = 60;
+
+   private ConcurrentHashMap<String, CompletableFuture<DeployedCacheStoreMetadata>> deployedCacheStores = new ConcurrentHashMap<>();
 
    @Override
    public <T> T createInstance(StoreConfiguration cfg) {
-      if(cfg instanceof DeployedStoreConfiguration) {
+      if (cfg instanceof DeployedStoreConfiguration) {
          DeployedStoreConfiguration deployedConfiguration = (DeployedStoreConfiguration) cfg;
-         DeployedCacheStoreMetadata deployedCacheStoreMetadata = deployedCacheStores.get(deployedConfiguration.getCustomStoreClassName());
-         return (T) deployedCacheStoreMetadata.getLoaderWriterRawInstance();
+         DeployedCacheStoreMetadata deployedCacheStoreMetadata = null;
+         try {
+            deployedCacheStoreMetadata = getPromise(deployedConfiguration.getCustomStoreClassName()).get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            return (T) deployedCacheStoreMetadata.getLoaderWriterRawInstance();
+         } catch (InterruptedException e) {
+            InfinispanLogger.ROOT_LOGGER.error("Could not get Custom Cache Store metadata (" + deployedConfiguration.getCustomStoreClassName() + ") because of an InterruptionException.", e);
```

Comment:
This part is invoked by server integration code... It doesn't have much asynchronous code which will catch 'InterruptedException'. So I'm not sure if setting interruped flag back will give us anything... On the other hand rethrowing it as a `RuntimeException` will prevent server deployer from further processing. 

I'm totally not sure what's the proper approach. Right now it will simply log an error and move on. I believe this is the best we can do at this point...

@tristantarrant any thoughts?